src/RuleSet/AutomaticMigrationSetTrait.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\RuleSet;
+
+use Composer\Semver\Semver;
+use PhpCsFixer\ComposerJsonReader;
+use PhpCsFixer\ConfigurationException\UnresolvableAutoRuleSetConfigurationException;
+
+/**
+ * @internal
+ *
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * @no-named-arguments Parameter names are not covered by the backward compatibility promise.
+ */
+trait AutomaticMigrationSetTrait
+{
+    private function calculateTargetSet(string $setName, string $entity, bool $isRisky): string
+    {
+        static $set = null;
+
+        if (null === $set) {
+            $actualVersion = self::calculateActualVersion($setName, $entity);
+
+            $candidates = self::calculateCandidateSets($entity, $isRisky);
+            $composerCandidates = Semver::rsort(array_keys($candidates));
+
+            foreach ($composerCandidates as $candidate) {
+                if (Semver::satisfies($actualVersion, '>='.$candidate)) {
+                    $set = $candidates[$candidate]; // @phpstan-ignore offsetAccess.notFound
+
+                    break;
+                }
+            }
+
+            if (null === $set) {
+                throw new UnresolvableAutoRuleSetConfigurationException(\sprintf('No migration set found feasible for %s %s.', $entity, $actualVersion));
+            }
+        }
+
+        return $set;
+    }
+
+    /**
+     * @return list<AbstractMigrationSetDescription>
+     */
+    private static function getMigrationSets(): array
+    {
+        static $sets = null;
+
+        if (null === $sets) {
+            $sets = array_filter(
```

Comment:
`$sets` is not a list because of `array_filter`. Phpstan will complain about it when the trait is used (traits are not analyzed standalone).

Reply:
c31e546112dfe28226807f4e2da79bd640920c39

but OK to fix complaints in #9068